File: DynamicPrograming/java/NonAdjacentSum.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class NonAdjacentSum {
  private NonAdjacentSum() {
  }

  public static int maxSum(List<Integer> nums) {
    Map<Integer, Integer> memo = new HashMap<>();
    return dfs(0, nums, memo);
  }

  private static int dfs(int index, List<Integer> nums, Map<Integer, Integer> memo) {
    if (index >= nums.size()) {
      return 0;
    }
    if (memo.containsKey(index)) {
      return memo.get(index);
    }
    int include = nums.get(index) + dfs(index + 2, nums, memo);
    int exclude = dfs(index + 1, nums, memo);
    int best = Math.max(include, exclude);
    memo.put(index, best);
    return best;
  }
}

File: DynamicPrograming/ts/nonAdjacentSum.ts
export function nonAdjacentSum(nums: number[]): number {
  const memo = new Map<number, number>();

  const dfs = (i: number): number => {
    if (i >= nums.length) return 0;
    if (memo.has(i)) return memo.get(i)!;

    const include = nums[i] + dfs(i + 2);
    const exclude = dfs(i + 1);
    const best = Math.max(include, exclude);
    memo.set(i, best);
    return best;
  };

  return dfs(0);
}
